pipe binary data to external commands

henryjetmundsen opened the pull request:
Fixes #7615 

# Description

When calling external commands, we create a table from the pipeline data to handle external commands expecting paginated input. When a binary value is made into a table, we convert the vector of bytes representing the binary bytes into a pretty formatted string. This results in the pretty formatted string being sent to external commands instead of the actual binary bytes. By checking whether the stdout of the call is being redirected, we can decide whether to send the raw binary bytes or the pretty formatted output when creating a table command.

# User-Facing Changes

When passing binary values to external commands, the external command will receive the actual bytes instead of the pretty printed string. Use cases that don't involve piping a binary value into an external command are unchanged.

![new_behavior](https://user-images.githubusercontent.com/32406734/218349172-24cd12f2-d563-4957-bdf1-6aa804b174b2.png)

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

    cargo fmt --all -- --check to check standard code formatting (cargo fmt --all applies these changes)
    cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect to check that you're using the standard code style
    cargo test --workspace to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

henryjetmundsen commented:
My changes have caused the [const_binary](https://github.com/nushell/nushell/blob/main/tests/const_/mod.rs#L30-L37) test to fail. The test is running `const x = 0x[12]; $x` and then verifying that `12` is in the output. You can see from my screenshot that this is still the actual behavior after my changes. However the test is failing, likely because `call.redirect_stdout` is set to `true` in tests but `false` in the repl. If the tests aren't indicative of real behavior, my intuition is that the tests need to be reworked to fix this. However, I'm unsure of the complexity involved in adjusting tests this way versus finding a different solution to this problem that doesn't use `call.redirect_stdout` to make decisions.

sholderbach commented on crates/nu-command/tests/commands/run_external.rs:
```diff
@@ -213,6 +213,17 @@ fn external_command_not_expand_tilde_with_quotes() {
+        let actual = nu!(cwd: dirs.test(), pipeline(r#"0x[deadbeef] | hexdump -C"#));
+        assert_eq!(
+            actual.out,
+            r#"00000000  de ad be ef                                       |....|00000004"#
```
I don't know if the padding of `hexdump`'s output is stable and set in stone, so this might make this test less stable. Also I am not sure if it is available by default on Windows systems were our tests are also run. I wouldn't be a fan of completely disabling the test on Windows as your PR introduces semantics that should be general.

henryjetmundsen replied:
Yeah I wasn't a fan of using hexdump in this test either, but I wasn't able to figure another way. I'm not sure of any default external program that can process hex bytes. Piping into `^cat` was outputting different bytes than it received so I didn't think it'd be stable to test that either.

WindSoilder replied:
What about removing padding stuff in `hexdump` to make our test reliable.

Generally like this:
```rust
assert_eq!(
    actual.out.replace(" ", ""),
   r#"00000000deadbeef|....|00000004"#
)
```

WindSoilder commented:
I get the reason:
https://github.com/nushell/nushell/blob/d5ce509e3afff0c7e04855c252bbacdcba381060/crates/nu-cli/src/eval_file.rs#L157-L162

Because we're rendering binary output based on `call.redirect_output` value, we need to adjust the final call on `table` command.

```rust
// The final call on table command, it's ok to set redirect_output to false.
let mut call = Call::new(Span::new(0, 0));
call.redirect_stdout = false;
let table = command.run(engine_state, stack, &call, pipeline_data);
```

Because it's the final call, it's ok

henryjetmundsen commented:
@WindSoilder I've updated the hexdump test to remove whitespace and followed your other suggestion for redirect_stdout. As @sholderbach mentioned, the hexdump test might still not pass on Windows and honestly I'm not sure if hexdump is on linux by default either. If the hexdump test doesn't work in CI/CD, I'm not sure of what default command could actually be used.

sholderbach commented:
As a shim for tests were we would have to assume too much about the system we have the `nu --testbins`. Maybe we could use one of the existing ones or add one for what you need (e.g. counting bytes on stdin)
The only thing I am not sure is if we have some special casing for `nu` that would interfere with the require test.

WindSoilder commented:
I found a way to test it.  Since it's mainly about `table` command and redirection, we can test it like this:
```nu
0x[abcdef] | table | bytes length
```

Before this pr, `bytes length` gives us error because `table` output string, but in this pr, it should return 3.

henryjetmundsen commented:
Great suggestion @WindSoilder! I think my changes now should be passing CI.

sholderbach commented:
Sorry to disagree with @WindSoilder on this one:
> Since it's mainly about `table` command and redirection, we can test it like this:

The test like this validates your current solution but not necessarily the semantics we want to have with *external* commands.

> ```nu
> 0x[abcdef] | table | bytes length
> ```

It only uses internal commands and could silently break in functionality with a change in other areas.
Furthermore the main problem described in #7615 is that of external commands in any pipeline position receiving `Value::Binary` from the pipeline. That `table` is load bearing is mostly just a description of the current state.

I would argue that an explicit call to `table` by the user would signify that they want the output at this step to be pretty-formatted based on the flags given to that. So I would even go as far that maybe users expect that after a `... | table | ...` you could expect a text stream or string.

henryjetmundsen commented:
I agree there is potential for silent failure if at some point in the future we

1. Don't turn the input into a table before passing to an external command
2. The new way we pass binary values to external commands happens to also be code that converts it to a pretty formatted string

However, I'm not sure I agree that if a user passes a binary value to `table` the expectation is that they get a pretty formatted string. The only places where `pretty_hex` are ever called is when a binary value is converted to a table or converted to html. The knowledge that we convert binary values to a table before returning them is an implementation detail that the end user shouldn't know about.

I believe that the expectation today is that if a Binary value is returned by a nu script, its value will be printed in pretty format. The end user doesn't know when they type `nu -c 0x[deadbeef]` that what's actually happening is that a `Vec<u8>` is being made into a table behind the scenes. To the end user, the pretty string is just how nu prints binary values to the screen. I also think the expectation today is that if a binary value is passed through a pipe, the command on the other side of the pipe should receive the binary value and not the pretty string. The reason why this is an issue specifically for external commands is because we convert the binary value to a table before going to an external command for pagination reasons.

So these changes do mean that a user is not able to perform operations on a pretty hex string. I also don't think there's a valid use case where someone would want to do that. I think the pretty hex string should exist solely as a way to present a binary value to a user when it's the final output from a nu script.

WindSoilder commented:
Hi, @henryjetmundsen sorry about it, we have discussed internally, we love the change.

But one more thing we can improve is about test case.   For the quality control, we thought that an ideal case would be making a new testbin like `--testbin input-length`.  You can check `src/test_bins.rs` and `src/main.rs` to see how to create a testbin.

So our code can be tested like `0x[abcdef] | nu --testbin input-length`(the result should be 3), it hides implementation detail.  Would you please adjust the test case?  If you feel into trouble to implement it, feel free to reach, I can help to push changes to your branch.

henryjetmundsen commented:
@WindSoilder sounds good! I've removed the test that was added to `run_external` and added a test bin `input_bytes_length` to check the number of bytes received from stdin.

WindSoilder commented:
Thanks!  Looks good to me

sholderbach commented on crates/nu-command/tests/commands/run_external.rs:
```diff
@@ -213,14 +213,6 @@ fn external_command_not_expand_tilde_with_quotes() {
-        let actual = nu!(cwd: dirs.test(), pipeline(r#"0x[deadbeef] | table | bytes length"#));
-        assert_eq!(actual.out, r#"4"#);
-    })
-}
```
Thanks for adding the testing utility!

Let's measure the `deadbeef` with your new testbin here :)

henryjetmundsen replied:
@sholderbach makes sense. Didn't realize we could call testbins that way. I've added the test back as `0x[deadbeef] | nu --testbin input_bytes_length`

sholderbach replied:
Awesome thanks, all tests seem to go green!

sholderbach commented:
Added a breaking change label as it changes the pipeline semantics for `Value::Binary`

sholderbach commented:
Ah sorry your PR got under the wheels of the release. I think we can land this. Thanks for making a great change and being active in the review process!

fdncred commented:
I just discovered that this PR broke our pretty-hex output. We may need to revert. We need to be able to do this
```
open -r Cargo.toml | into binary | take 100 
Length: 100 (0x64) bytes | printable whitespace ascii_other non_ascii
00000000:   5b 70 61 63  6b 61 67 65  5d 0a 61 75  74 68 6f 72   [package]_author
00000010:   73 20 3d 20  5b 22 54 68  65 20 4e 75  73 68 65 6c   s = ["The Nushel
00000020:   6c 20 50 72  6f 6a 65 63  74 20 44 65  76 65 6c 6f   l Project Develo
00000030:   70 65 72 73  22 5d 0a 64  65 66 61 75  6c 74 2d 72   pers"]_default-r
00000040:   75 6e 20 3d  20 22 6e 75  22 0a 64 65  73 63 72 69   un = "nu"_descri
00000050:   70 74 69 6f  6e 20 3d 20  22 41 20 6e  65 77 20 74   ption = "A new t
00000060:   79 70 65 20                                          ype
```
@henryjetmundsen Any ideas how to fix this?

fdncred commented:
I've discovered two work-arounds.

1. `open -r Cargo.toml | into binary | take 100 | table`
2. change the `$env.config.hooks.display_output` hook to `table`, so it always calls the `table` command.
